db/view/view.cc

```diff
@@ -2682,11 +2682,69 @@ static future<> flush_base(lw_shared_ptr<replica::column_family> base, abort_sou
     }).discard_result();
 }
 
+void view_builder::update_base_schema(const replica::column_family& base) {
+    auto base_schema = base.schema();
+    auto& views = base.views();
+    if (views.empty()) {
+        return;
+    }
+    auto step_it = _base_to_build_step.find(base_schema->id());
+    if (step_it == _base_to_build_step.end()) {
+        return;// In case all the views for this CF have finished building already.
+    }
+    if (step_it->second.base_schema->version() == base_schema->version()) {
+        return;// Base schema was already updated, probably while creating an index.
+    }
+    step_it->second.base_schema = base_schema;
+    // Update view_ptrs in all build statuses with new view_ptr from the base table.
+    // If some were dropped, stop them as well, so we won't continue building them,
+    // especially as their schema is not updated with the new base schema.
+    // If some were added, we can handle them later, in on_create_view.
+    std::vector<view_ptr> views_to_remove;
+    for (auto& status : step_it->second.build_status) {
+        auto view_it = std::ranges::find_if(views, [&status] (const view_ptr& v) {
+            return v->id() == status.view->id();
+        });
+        if (view_it == views.end()) {
+            views_to_remove.push_back(status.view);
+        } else {
+            // This update will be performed again in on_update_view, but it will set the same
+            // view_ptr, and it's still needed there because not all view schema updates are
+            // caused by base schema updates.
+            status.view = *view_it;
+        }
+    }
+    for (auto& view : views_to_remove) {
+        _built_views.erase(view->id());
+        step_it->second.build_status.erase(std::ranges::find_if(step_it->second.build_status, [&view] (const view_build_status& bs) {
+            return bs.view->id() == view->id();
+        }));
+        // The cleanup operations will be done in on_drop_view.
```

Comment:
I'm not sure that I like this. This potentially puts the view_builder into a weird state where some cleanup has already been made for a view that were already removed, but we still rely on `on_drop_view` to finish the job.

Because we are now forced to update information about more than one schema object (i.e. views and tables), I think a cleaner approach would be to have `update_base_schema` and `update_view_schema` functions which synchronizes information inside `_base_to_build_step` related to the thing being changed with the status quo. `on_create_view`/`on_update_view`/`on_drop_view` would just call it.

Reply:
I modified this in the rebase, all the drop operations are now done during `on_update_column_family`. For now I only extracted the view dropping part. I'm not sure if a general `update_view_schema` would simplify things that much - it would need to have all the logic of `on_update_view` and `on_drop_view`, some extra to differentiate them (we lookup the schema differently in both cases) and we would still need to process the view schemas in `update_base_schema` similarly to avoid dropping a view while iterating over all views in build step